### PR TITLE
Add concept `random` and concept exercise `maze-maker`

### LINKED ISSUE #537: Concept: Random
# Design

## Goal

The Goal is to learn to use the `Random` module.

## Learning objectives

Students should be able to
- understand the constraints of a pure language when it comes to generating randomness (mention `step`, `Seed`)
- understand what a generator is
- create primitive generators (`int`, `float`, `constant`)
- pick a distribution (`uniform`, `weighted`)
- generate data structures (`list`, `pair`)
- modify and intricate generators (`map`, `andThen`)
- define recursive generators (`lazy`)

## Out of scope

- `generate`, `Platform` and `Cmd msg` (a quick mention is fine)
- details on the algorithm, `minInt` and `maxInt`

## Concepts

The concept this exercise unlocks is:

- `random`

## Prerequisites

- basics-1
- basics-2
- lists
- custom types

## Analyzer

TBD depending on the exercise

-----------

### Comments

- The exercise should go beyond the usual "pick a thing from a list" and highlight the power of generators, for example building a recursive data structure (maze? family tree? Elm program?...)
- Test the exercise using [seed and steps](https://package.elm-lang.org/packages/elm/random/latest/Random#Seed), in such a manner that students can also use it to debug their code.
- mention [elm-community/random-extra](https://package.elm-lang.org/packages/elm-community/random-extra/latest) has a lot of useful resources

### PR DESCRIPTION
**Pull request by jiegillet**
Closes #537 

I had been thinking about doing this one for a while.
~~This is actually blocked by #549 (first time using fuzz tests) and https://github.com/exercism/elm-test-runner/pull/49 (parsing fuzztests)~~, but since I think there won't be big problems there and because I was motivated, I started working on this.

@pwadsworth if you are so inclined, I would love a review from you. If you could try to solve the exercise without looking at the exemplar and share your solution and impressions (is it too hard? too easy?) that would be great. As well as any other comment you might have of course.

The CI is expected to fail for two reasons: ~~the test-runner will not be able to parse the exercise~~ (EDIT: that's taken care of), and the introduction is missing from the exercise, I will copy it once it's been reviewed.

### CODE REVIEW

File: concepts/random/introduction.md
Comment (jiegillet) [inline concepts/random/introduction.md:1]: I decided to keep the intro and about almost exactly the same, because pretty much everything is needed for solving the exercise.
The only change I did is remove all the links, because they are considered distracting.

File: exercises/concept/maze-maker/tests/Tests.elm
Comment (jiegillet) [inline exercises/concept/maze-maker/tests/Tests.elm:123]: The probability to get a depth of 3 is the smallest, just about 3%. Doing 1000 runs gives a <99.99% chance of that happening.

**Review by pwadsworth — COMMENTED** [review]
The directions are clear and it helps that the requirements follow the same order as the concept's `about.md`. Here is my solution without looking at the exemplar. I did have to look at the hints in task 4 because the need to flatten the result, and how to do it, with `Random.andThen identity` were not obvious to me. 

```elm
module MazeMaker exposing (..)

import Random exposing (Generator)


type Maze
    = DeadEnd
    | Room Treasure
    | Branch (List Maze)


type Treasure
    = Gold
    | Diamond
    | Friendship


deadend : Generator Maze
deadend =
    Random.constant DeadEnd


treasure : Generator Treasure
treasure =
    Random.uniform Gold [ Diamond, Friendship ]


room : Generator Maze
room =
    Random.map Room treasure


branch : Generator Maze -> Generator Maze
branch mazeGenerator =
    Random.int 2 4
        |> Random.andThen (\n -> Random.list n mazeGenerator)
        |> Random.map Branch


maze : Generator Maze
maze =
    Random.weighted ( 60, deadend )
        [ ( 15, room )
        , ( 25, branch (Random.lazy (\_ -> maze)) )
        ]
        |> Random.andThen identity


mazeOfDepth : Int -> Generator Maze
mazeOfDepth depth =
    if depth == 0 then
        Random.uniform deadend [ room ]
            |> Random.andThen identity

    else
        branch (Random.lazy (\_ -> mazeOfDepth (depth - 1)))
```

File: concepts/random/about.md
Comment (jiegillet) [reply concepts/random/about.md:121]: Well, I did explain what `Random.andThen` does, and I think it's pretty clear what `identity` is.
But really, I don't want to explain this too much, because I feel it's the only "challenging" part which requires a bit of thought. And it's not really that challenging either since I show it here used in exactly the same way and also I left a hint.
I'll think about it some more :)

File: concepts/random/about.md
Comment (jiegillet) [reply concepts/random/about.md:9]: OK, I'll add a section about it in the about.md, but I'll keep it super short for the introduction.

**Comment by ceddlyburge** [comment]
Hi @jiegillet , I'm back from holiday now, do you want me to take a look at this?

**Comment by jiegillet** [comment]
Welcome back!
Yes, eventually I want your opinion, but we should tackle the blocking PRs first.

File: concepts/random/about.md
Comment (ceddlyburge) [inline concepts/random/about.md:44]:
```suggestion
From now on, we will focus on generators.
```

File: concepts/random/about.md
Comment (ceddlyburge) [inline concepts/random/about.md:56]:
```suggestion
Those values can be combined into lists of values, or tuples / pairs:
```

File: concepts/random/about.md
Comment (ceddlyburge) [inline concepts/random/about.md:82]:
```suggestion
`Random.uniform` takes two arguments (`Red, [Green, Blue]`) to guarantee that there is at least one value to pick from.
Otherwise if it took one list argument (`[Red, Green, Blue`]) then the list could be empty.
```

File: concepts/random/about.md
Comment (ceddlyburge) [reply concepts/random/about.md:149]: I find this code pretty confusing as well, I can't say I understand at the moment. I do understand identity, and andThen, so maybe its lazy that I am finding confusing. From the comment showing the output, it looks like `generate 3` will create 3 uniformly distributed natural numbers between 0 and 2, is that right? And if so, what controls the range of the natural numbers produced?

File: exercises/concept/maze-maker/.docs/hints.md
Comment (ceddlyburge) [inline exercises/concept/maze-maker/.docs/hints.md:32]:
```suggestion
- Use the [`Random.uniform`][random-uniform] function for generating the dead ends or rooms at depth `0`
```

This looks like a typo to me, but not sure exactly what the wording should be

File: exercises/concept/maze-maker/.docs/hints.md
Comment (ceddlyburge) [inline exercises/concept/maze-maker/.docs/hints.md:33]: Maybe make this the same as the other similar comment, and mention `identity`.

File: exercises/concept/maze-maker/.meta/design.md
Comment (ceddlyburge) [inline exercises/concept/maze-maker/.meta/design.md:24]:
```suggestion
- `maxInt`, `minInt`
```

File: exercises/concept/maze-maker/src/MazeMakerSupport.elm
Comment (ceddlyburge) [inline exercises/concept/maze-maker/src/MazeMakerSupport.elm:1]: I don't think these are used in the exemplar / solution, so maybe its better for these to go in the test module? Maybe I'm missing something ...

File: exercises/concept/maze-maker/tests/Tests.elm
Comment (ceddlyburge) [reply exercises/concept/maze-maker/tests/Tests.elm:123]: Maybe it is enough to test to depth 2 here?

**Review by ceddlyburge — APPROVED** [review]
Looks good, Random is a complex topic! It will be a great addition to the track.
I've left a few comments but nothing major.

File: concepts/random/about.md
Comment (jiegillet) [reply concepts/random/about.md:149]: Maybe this example is too strange? I was trying to find one that's simple enough and different enough from the exercise one. Let me know if you have a better idea.

First of all, `lazy`doesn't add anything to the logic, all it does is wrap the function in a lambda so that the compiler doesn't infinitely stick `peano` into `peano`.

Then, what `peano` does is return `Zero` with a 50% chance , and itself plus one the other 50%. So the numbers coming out should be 0 (50% of the time), 1 (25% of the time), 2 (12% of the time), 3 (6% of the time), etc. The numbers are not bounded, because the generator could keep picking `peano` an arbitrary number of time, but it's more and more unlikely.

Maybe I should generate more numbers to get a better feel for the distribution.

File: exercises/concept/maze-maker/src/MazeMakerSupport.elm
Comment (jiegillet) [reply exercises/concept/maze-maker/src/MazeMakerSupport.elm:1]: I do mention `MazeMakerSupport.mazeDepth` in the instructions, so I thought it would be nice to show it. 
Also, if I put those functions in the test file, the students will not be able to look at the definition at the same time as their failing test.
In general, I find that I almost never look at the test file since Exercism shows me the test snippet, so maybe this is more discoverable? I don't really know 🤷‍♂️ 

What do you think?

File: exercises/concept/maze-maker/tests/Tests.elm
Comment (jiegillet) [reply exercises/concept/maze-maker/tests/Tests.elm:123]: I mention the probability to get depth 3 in the next task, so I thought it would be nice to actually get it.

I thought 2 was a bit underwhelming, 1000 runs is really fast anyway, and a <0.01% chance of failure is reasonable.

File: concepts/random/about.md
Comment (ceddlyburge) [reply concepts/random/about.md:149]: Ah I see, that does make sense, and this explanation is great :)

Maybe add it to the document?

File: exercises/concept/maze-maker/src/MazeMakerSupport.elm
Comment (ceddlyburge) [reply exercises/concept/maze-maker/src/MazeMakerSupport.elm:1]: I don't think its a big deal either way, we could just wait for any feedback?

The only downside I see is that students might think they have to use it, being as its provided. I have no idea if anyone will think this though ...

File: concepts/random/about.md
Comment (jiegillet) [reply concepts/random/about.md:149]: Done in [2b0732d](https://github.com/exercism/elm/pull/613/commits/2b0732de91c18562b7d9205a3f54202ff71ae382)

**Comment by jiegillet** [comment]
Ok, I think I've addressed all the comments. I'll merge it after the weekend in case someone wants to take a closer look.

**Comment by ceddlyburge** [comment]
Looks good to me ...

**Pull request merged**
